webapi/src/main/twirl/queries/sparql/v1/createNewUser.scala.txt

```diff
@@ -0,0 +1,94 @@
+ * This file is part of Knora.
+ *
+ * Knora is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Knora is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with Knora.  If not, see <http://www.gnu.org/licenses/>.
+ *@
+
+@import org.knora.webapi.IRI
+
+@*
+ * Creates a new user.
+ *
+ * @param triplestore the name of the triplestore being used. The template uses this value to exclude inferred
+                      results from the WHERE clause of the update.
+ * @param userIri the Iri of the new user
+ * @param userClassIri the IRI of the OWL class that the new user should belong to.
+ * @param email the new user's email address.
+ * @param password the new user's password.
+ * @param givenName the new user's given name.
+ * @param familyName the new user's family name.
+ * @param status the new user's status (active = true, inactive = false)
+ * @param preferredLanguage the new user's preferred language.
+ * @param systemAdmin the new user's system admin group membership.
+ *
+ *@
+@(adminNamedGraphIri: IRI,
+  triplestore: String,
+  userIri: IRI,
+  userClassIri: IRI,
+  email: String,
+  password: String,
+  givenName: String,
+  familyName: String,
+  status: Boolean,
+  preferredLanguage: String,
+  systemAdmin: Boolean)
+
+prefix xsd: <http://www.w3.org/2001/XMLSchema#>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+PREFIX knora-base: <http://www.knora.org/ontology/knora-base#>
+
+INSERT {
+    GRAPH ?adminNamedGraphIri {
+        ?userIri rdf:type ?userClassIri .
+
+        ?userIri knora-base:email "@email"^^xsd:string .
+
+        ?userIri knora-base:password "@password"^^xsd:string .
+
+        ?userIri knora-base:givenName "@givenName"^^xsd:string .
+
+        ?userIri knora-base:familyName "@familyName"^^xsd:string .
+
+        ?userIri knora-base:status "@status"^^xsd:boolean .
+
+        ?userIri knora-base:preferredLanguage "@preferredLanguage"^^xsd:string .
+
+        ?userIri knora-base:isInSystemAdminGroup "@systemAdmin"^^xsd:boolean .
+    }
+}
+@*
+
+GraphDB's consistency checking requires reasoning, but reasoning interferes with certain things
+in the WHERE clauses of our SPARQL updates, so we set a GraphDB-specific flag to return only
+explicit statements in the WHERE clause here.
+
+*@
+@triplestore match {
+    case "graphdb" | "graphdb-free" => {
+        USING <http://www.ontotext.com/explicit>
+    }
+
+    case other => {}
+}
+WHERE {
```

Comment:
> The Scala code first does a check to see if the email is already used, before the ``createNewUser`` update query is executed. 

Yes, but there is some time between that check and the update query. The same user could be created by another thread during that time.

> If there is a global lock on ``http://data.knora.org/users`` as you suggested, then ``FILTER NOT EXISTS`` is not really needed.

I think redundant checks are a good thing. If one check fails for some reason (e.g. because of a bug), the others might still work.

Please read the sections "Application-Level Locking" and "Ensuring Data Consistency" here:

http://www.knora.org/documentation/manual/rst/knora-api-server/design-documentation/triplestore-updates.html#application-level-locking